packages/terminal/src/browser/terminal-frontend-contribution.ts

```diff
@@ -205,11 +206,38 @@ export class TerminalFrontendContribution implements FrontendApplicationContribu
         });
     }
 
-    onWillStop(): boolean {
-        return (
-            this.terminalPreferences['terminal.integrated.confirmOnExit'] !== 'never' &&
-            this.widgetManager.getWidgets(TERMINAL_WIDGET_FACTORY_ID).length > 0
-        );
+    onWillStop(): OnWillStopAction | undefined {
+        const preferenceValue = this.terminalPreferences['terminal.integrated.confirmOnExit'];
+        if (preferenceValue !== 'never') {
+            const allTerminals = this.widgetManager.getWidgets(TERMINAL_WIDGET_FACTORY_ID) as TerminalWidget[];
+            if (allTerminals.length) {
+                return {
+                    action: async () => {
+                        if (preferenceValue === 'always') {
+                            return this.confirmExitWithActiveTerminals(allTerminals.length);
+                        } else {
+                            const activeTerminals = await Promise.all(allTerminals.map(widget => widget.hasChildProcesses()))
+                                .then(hasChildProcesses => hasChildProcesses.filter(hasChild => hasChild));
+                            return activeTerminals.length === 0 || this.confirmExitWithActiveTerminals(activeTerminals.length);
+                        }
+                    },
+                    reason: 'Active integrated terminal',
+                };
+            }
+        }
+    }
+
+    protected async confirmExitWithActiveTerminals(activeTerminalCount: number): Promise<boolean> {
+        const msg = activeTerminalCount === 1
+            ? nls.localizeByDefault('Do you want to terminate the active terminal session?')
+            : nls.localizeByDefault('Do you want to terminate the {0} active terminal sessions?', activeTerminalCount);
+        const safeToExit = await new ConfirmDialog({
+            title: '',
+            msg,
+            ok: nls.localizeByDefault('Terminate'),
```

Comment:
I mentioned earlier how I would like to see this addressed. Afterwards, I think this would be good to go 👍 

```suggestion
            ? nls.localize('theia/terminal/terminateActive', 'Do you want to terminate the active terminal session?')
            : nls.localize('theia/terminal/terminateActiveMultiple', 'Do you want to terminate the {0} active terminal sessions?', activeTerminalCount);
        const safeToExit = await new ConfirmDialog({
            title: '',
            msg,
            ok: nls.localize('theia/terminal/terminate', 'Terminate'),
```